bitswap.go

```diff
@@ -325,56 +332,91 @@ func (bs *Bitswap) ReceiveMessage(ctx context.Context, p peer.ID, incoming bsmsg
 		return
 	}
 
-	wg := sync.WaitGroup{}
-	for _, block := range iblocks {
-
-		wg.Add(1)
-		go func(b blocks.Block) { // TODO: this probably doesnt need to be a goroutine...
-			defer wg.Done()
-
-			bs.updateReceiveCounters(b)
-			bs.sm.UpdateReceiveCounters(p, b)
-			log.Debugf("[recv] block; cid=%s, peer=%s", b.Cid(), p)
-			// skip received blocks that are not in the wantlist
-			if !bs.wm.IsWanted(b.Cid()) {
-				log.Debugf("[recv] block not in wantlist; cid=%s, peer=%s", b.Cid(), p)
-				return
-			}
+	bs.updateReceiveCounters(iblocks)
+	for _, b := range iblocks {
+		log.Debugf("[recv] block; cid=%s, peer=%s", b.Cid(), p)
+	}
 
-			if err := bs.receiveBlockFrom(b, p); err != nil {
-				log.Warningf("ReceiveMessage recvBlockFrom error: %s", err)
-			}
-			log.Event(ctx, "Bitswap.GetBlockRequest.End", b.Cid())
-		}(block)
+	// Split blocks into wanted blocks vs duplicates
+	wanted := make([]blocks.Block, 0, len(iblocks))
+	var dups []blocks.Block
+	for _, b := range iblocks {
+		if bs.wm.IsWanted(b.Cid()) {
+			wanted = append(wanted, b)
+		} else {
+			dups = append(dups, b)
+			log.Debugf("[recv] block not in wantlist; cid=%s, peer=%s", b.Cid(), p)
+		}
 	}
-	wg.Wait()
-}
 
-func (bs *Bitswap) updateReceiveCounters(b blocks.Block) {
-	blkLen := len(b.RawData())
-	has, err := bs.blockstore.Has(b.Cid())
+	// Process blocks
+	err := bs.receiveBlocksFrom(p, wanted, dups, true)
 	if err != nil {
-		log.Infof("blockstore.Has error: %s", err)
+		log.Warningf("ReceiveMessage recvBlockFrom error: %s", err)
 		return
 	}
 
-	bs.allMetric.Observe(float64(blkLen))
-	if has {
-		bs.dupMetric.Observe(float64(blkLen))
+	for _, b := range wanted {
+		log.Event(ctx, "Bitswap.GetBlockRequest.End", b.Cid())
 	}
+}
+
+func (bs *Bitswap) updateReceiveCounters(blocks []blocks.Block) {
+	// Check which blocks are in the datastore
+	blocksHas := bs.blockstoreHas(blocks)
 
 	bs.counterLk.Lock()
 	defer bs.counterLk.Unlock()
-	c := bs.counters
 
-	c.blocksRecvd++
-	c.dataRecvd += uint64(len(b.RawData()))
-	if has {
-		c.dupBlocksRecvd++
-		c.dupDataRecvd += uint64(blkLen)
+	// Do some accounting for each block
+	for i, b := range blocks {
+		has := blocksHas[i]
+		if has == nil {
+			// If there was an error checking the datastore for a block CID,
+			// the error was logged and the CID was not added to the map
+			return
```

Comment:
We probably shouldn't _ignore_ the other results on error.

Reply:
Yeah that's true, should we just assume they're uniques?

Reply:
By unique do you mean one-off errors? Yeah, we can get random errors e.g., due to out of fd conditions (although we need to work around those better).

Reply:
I meant that any time a call to `blockstore.Has()` fails, let's assume the block is not in the blockstore; ie the block is a unique, as opposed to being a duplicate.